### PR TITLE
Fix race which can lead to merged parts premature removal

### LINKED ISSUE #43983: No such file or directory: `data.cmrk3`
https://s3.amazonaws.com/clickhouse-test-reports/0/39bda355f6a41b16bb015d9039884eca6508be6f/stateless_tests__asan__[2/2].html

https://s3.amazonaws.com/clickhouse-test-reports/43749/a50482ecd34b68e65566d5fdc4ed895a97e5ceb3/stateless_tests__tsan__[1/5].html

https://s3.amazonaws.com/clickhouse-test-reports/43686/2b6661f87c90e9c398e2babb5c366b3826d9a038/stateless_tests__ubsan__[2/2].html
```
2022-12-06 04:42:32 00993_system_parts_race_condition_drop_zookeeper:                       [ FAIL ] 62.48 sec. - return code: 233
2022-12-06 04:42:32 [62141c37099b] 2022.12.06 04:42:31.779473 [ 160952 ] {91ab5e5d-7591-4925-aba2-9627f4c369a3} <Error> executeQuery: std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/31e/31eb607c-a552-44f7-8f02-4c1bc694f2fa/-1_0_221_5_247/data.cmrk3"]
2022-12-06 04:42:32 Cannot print extra info for Poco::Exception (version 22.12.1.1 (official build)) (from [::1]:42654) (comment: 00993_system_parts_race_condition_drop_zookeeper.sh) (in query: SELECT if((countDistinct(data) as c) == 0, 1, c) FROM ( SELECT _table, (count(), sum(a), sum(b), round(sum(c))) AS data FROM merge(currentDatabase(), 'alter_table_') GROUP BY _table )), Stack trace (when copying this message, always include the lines below):
2022-12-06 04:42:32 
2022-12-06 04:42:32 0. ./build_docker/../contrib/llvm-project/libcxx/include/exception:134: std::runtime_error::runtime_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) @ 0x43725a4d in /usr/bin/clickhouse
2022-12-06 04:42:32 1. ./build_docker/../contrib/llvm-project/libcxx/include/string:1499: std::__1::system_error::system_error(std::__1::error_code, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) @ 0x4374a3ee in /usr/bin/clickhouse
2022-12-06 04:42:32 2. ./build_docker/../contrib/llvm-project/libcxx/include/__filesystem/filesystem_error.h:42: std::__1::__fs::filesystem::filesystem_error::filesystem_error[abi:v15000](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::__fs::filesystem::path const&, std::__1::error_code) @ 0x1f7fd753 in /usr/bin/clickhouse
2022-12-06 04:42:32 3. ./build_docker/../contrib/llvm-project/libcxx/include/__filesystem/filesystem_error.h:90: void std::__1::__fs::filesystem::__throw_filesystem_error[abi:v15000]<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::__fs::filesystem::path const&, std::__1::error_code const&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::__fs::filesystem::path const&, std::__1::error_code const&) @ 0x435f3bfa in /usr/bin/clickhouse
2022-12-06 04:42:32 4. ./build_docker/../contrib/llvm-project/libcxx/src/filesystem/filesystem_common.h:173: std::__1::__fs::filesystem::detail::(anonymous namespace)::ErrorHandler<unsigned long>::report(std::__1::error_code const&) const @ 0x4360374e in /usr/bin/clickhouse
2022-12-06 04:42:32 5. ./build_docker/../contrib/llvm-project/libcxx/src/filesystem/operations.cpp:0: std::__1::__fs::filesystem::__file_size(std::__1::__fs::filesystem::path const&, std::__1::error_code*) @ 0x436032b6 in /usr/bin/clickhouse
2022-12-06 04:42:32 6. ./build_docker/../src/Disks/DiskLocal.cpp:0: DB::DiskLocal::getFileSize(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x2f1c20ba in /usr/bin/clickhouse
2022-12-06 04:42:32 7. ./build_docker/../src/Storages/MergeTree/DataPartStorageOnDisk.cpp:0: DB::DataPartStorageOnDisk::getFileSize(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x33765e97 in /usr/bin/clickhouse
2022-12-06 04:42:32 8. ./build_docker/../src/Storages/MergeTree/MergeTreeMarksLoader.cpp:0: DB::MergeTreeMarksLoader::loadMarksImpl() @ 0x33d1112d in /usr/bin/clickhouse
2022-12-06 04:42:32 9. ./build_docker/../contrib/llvm-project/libcxx/include/__utility/swap.h:37: DB::MergeTreeMarksLoader::loadMarks() @ 0x33d0f95a in /usr/bin/clickhouse
2022-12-06 04:42:32 10. ./build_docker/../src/Storages/MergeTree/MergeTreeMarksLoader.cpp:199: std::__1::__packaged_task_func<DB::MergeTreeMarksLoader::loadMarksAsync()::$_1, std::__1::allocator<DB::MergeTreeMarksLoader::loadMarksAsync()::$_1>, std::__1::shared_ptr<DB::MarksInCompressedFile> ()>::operator()() @ 0x33d12e42 in /usr/bin/clickhouse
2022-12-06 04:42:32 11. ./build_docker/../contrib/llvm-project/libcxx/include/future:0: std::__1::packaged_task<std::__1::shared_ptr<DB::MarksInCompressedFile> ()>::operator()() @ 0x33d18382 in /usr/bin/clickhouse
2022-12-06 04:42:32 12. ./build_docker/../base/base/wide_integer_impl.h:774: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x1f967d7c in /usr/bin/clickhouse
2022-12-06 04:42:32 13. ./build_docker/../src/Common/ThreadPool.h:197: ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x1f971870 in /usr/bin/clickhouse
2022-12-06 04:42:32 14. ./build_docker/../base/base/wide_integer_impl.h:774: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x1f9618cd in /usr/bin/clickhouse
2022-12-06 04:42:32 15. ./build_docker/../contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x1f96ba60 in /usr/bin/clickhouse
2022-12-06 04:42:32 16. ? @ 0x7f3642421609 in ?
2022-12-06 04:42:32 17. __clone @ 0x7f3642346133 in ?
```

Related to https://github.com/ClickHouse/ClickHouse/pull/43288 (I have not seen similar errors with uncompressed marks and it started after ~2022-11-24)

cc: @SmitaRKulkarni

### PR DESCRIPTION
**Pull request by alesapin**
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix race which can lead to premature temp parts removal before merge finished in ReplicatedMergeTree. This issue could lead to errors like `No such file or directory: xxx`. Fixes #43983.

We shouldn't destroy `MergeTask` before rename because it holds `scope_guard temporary_directory_lock;`. I don't think that we need a special comment here because destroy of task just looks redundant here.

Example log:
```
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
2022.12.05 20:11:07.640255 [ 802 ] {} <Trace> 0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199 (MergeFromLogEntryTask): Executing log entry to merge parts -1_0_0_0_199, -1_7_61_2_199, -1_67_67_0_199, -1_71_71_0_199, -1_81_81_0_199, -1_83_83_0_199, -1_88_88_0_199 to -1_0_88_3_199                                                                                                                                                                                                                                                                                                                                                                     
2022.12.05 20:11:08.591893 [ 798 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Debug> MergeTask::PrepareStage: Merging 7 parts: from -1_0_0_0_199 to -1_88_88_0_199 into Compact                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
2022.12.05 20:11:08.592773 [ 798 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Debug> MergeTask::PrepareStage: Selected MergeAlgorithm: Horizontal                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
2022.12.05 20:11:08.592969 [ 798 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Debug> MergeTreeSequentialSource: Reading 2 marks from part -1_0_0_0_199, total 2544 rows starting from the beginning of the part                                                                                                                                                                                                                                                                                                                                                                                                                                    
<... many similar messages ...>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
2022.12.05 20:11:11.930029 [ 926 ] {} <Warning> test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb): /var/lib/clickhouse/store/0f4/0f468411-a5b2-4bcd-af71-a4265d21c7eb/tmp_merge_-1_0_88_3_199/ is in use (by merge/mutation/INSERT) (consider increasing temporary_directories_lifetime setting)                                                                                                                                                                                                                                                                                                                                              
<... many similar messages ...>                                                                                                                             
2022.12.05 20:11:24.445675 [ 806 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Debug> MergeTask::MergeProjectionsStage: Merge sorted 39927 rows, containing 7 columns (7 merged, 0 gathered) in 16.804814082 sec., 2375.9263152316976 rows/sec., 129.33 KiB/sec.                                                                                                                                                               
2022.12.05 20:11:24.539378 [ 811 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Trace> MergedBlockOutputStream: filled checksums -1_0_88_3_199 (state Temporary)                                                                                                                                                                                                                                                                
2022.12.05 20:11:26.102672 [ 871 ] {} <Warning> test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb): /var/lib/clickhouse/store/0f4/0f468411-a5b2-4bcd-af71-a4265d21c7eb/tmp_merge_-1_0_88_3_199/ is in use (by merge/mutation/INSERT) (consider increasing temporary_directories_lifetime setting)                                                                                                                         
<... many similar messages ...>                                                                                                        
2022.12.05 20:11:27.027025 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Trace> test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb): Renaming temporary part tmp_merge_-1_0_88_3_199 to -1_0_88_3_199 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
2022.12.05 20:11:27.027399 [ 883 ] {} <Warning> test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb): Removing temporary directory /var/lib/clickhouse/store/0f4/0f468411-a5b2-4bcd-af71-a4265d21c7eb/tmp_merge_-1_0_88_3_199/
2022.12.05 20:11:27.027957 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Trace> test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb) (MergerMutator): Merged 7 parts: from -1_0_0_0_199 to -1_88_88_0_199
2022.12.05 20:11:27.035510 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Test> VersionMetadata: Trying to lock removal_tid by (1, 1, 00000000-0000-0000-0000-000000000000), table: test_thx6gyus.alter_table_6 (0f468411-a5b2-4bcd-af71-a4265d21c7eb), part: -1_0_0_0_199
<... many similar messages ...>      
2022.12.05 20:11:27.052695 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Test> test_thx6gyus.alter_table_6 (ReplicatedMergeTreeQueue): Removing successful entry queue-0000000640 from queue with type MERGE_PARTS with virtual parts [-1_0_88_3_199]
2022.12.05 20:11:27.052845 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Test> test_thx6gyus.alter_table_6 (ReplicatedMergeTreeQueue): Adding parts [-1_0_88_3_199] to current parts
2022.12.05 20:11:27.053534 [ 797 ] {0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199} <Test> test_thx6gyus.alter_table_6 (ReplicatedMergeTreeQueue): Removing part -1_0_88_3_199 from mutations (remove_part: false, remove_covered_parts: true)
2022.12.05 20:11:27.107046 [ 797 ] {} <Debug> MemoryTracker: Peak memory usage to apply mutate/merge in 0f468411-a5b2-4bcd-af71-a4265d21c7eb::-1_0_88_3_199: 8.68 MiB.

1278002:2022.12.05 20:11:54.027669 [ 53967 ] {502c49ba-6126-4229-a8df-018ecb9a39ae} <Error> executeQuery: std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/0f4/0f468411-a5b2-4bcd-af71-a4265d21c7eb/-1_0_88_3_199/data.cmrk3"]
```

### CODE REVIEW

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp [hunk 1]
```diff
@@ -297,8 +297,6 @@ bool MergeFromLogEntryTask::finalize(ReplicatedMergeMutateTaskBase::PartLogWrite
 {
     part = merge_task->getFuture().get();
 
-    /// Task is not needed
-    merge_task.reset();
     storage.merger_mutator.renameMergedTemporaryPart(part, parts, NO_TRANSACTION_PTR, *transaction_ptr);
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeFromLogEntryTask.cpp:300]: Should we reset it after renaming the part?

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp (hunk 1 above)
Comment (alesapin) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:300]: For what?

Comment (tavplubix) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:300]: I don't know, it was reset for some reason... Maybe @nikitamikhaylov remembers

**Comment by alesapin** [comment]
`test_distributed_queries_stress` -- I'd prefer to remove this test.

**Comment by alesapin** [comment]
Stress: one is missing blob, two others looks like OOM.

**Pull request merged**